bitcoin/src/witness_program.rs

```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
+                Ok(_) => assert_eq!(None, Some("Error, invalid program decoded successfully")),
+                Err(e) => {
+                    assert_eq!(WitnessProgramError::MismatchedProgramLength(18, 20), e)
+                }
+            }
+        }
+
+        #[test]
+        fn from_str_valid_p2sh_p2wpkh_program() {
+            match WitnessProgram::from_str(VALID_P2SH_P2WPKH_PROGRAM) {
+                Ok(witness_program) => {
+                    assert_eq!(0x00, witness_program.version);
+                    assert_eq!(vec![0x75, 0x1e, 0x76, 0xe8, 0x19,
+                                    0x91, 0x96, 0xd4, 0x54, 0x94,
+                                    0x1c, 0x45, 0xd1, 0xb3, 0xa3,
+                                    0x23, 0xf1, 0x43, 0x3b, 0xd6
+                    ], witness_program.program);
+                }
+                Err(e) => assert_eq!(None, Some(e))
+            }
+        }
+
+        const VALID_P2SH_P2WSH_PROGRAM: &str = "00201863143c14c5166804bd19203356da136c985678cd4d27a1b8c6329604903262";
+        const INVALID_P2SH_P2WSH_PROGRAM_LENGTH: &str = "00201863143c14c5166804bd19203356da136c985678cd4d27a1b8c632960490";
+
+        #[test]
+        fn from_str_invalid_p2sh_p2wsh_program_len() {
+            match WitnessProgram::from_str(INVALID_P2SH_P2WSH_PROGRAM_LENGTH) {
+                Ok(_) => assert_eq!(None, Some("Error, invalid program decoded successfully")),
+                Err(e) => {
+                    assert_eq!(WitnessProgramError::MismatchedProgramLength(30, 32), e)
+                }
+            }
+        }
+
+        #[test]
+        fn from_str_valid_p2sh_p2wsh_program() {
+            match WitnessProgram::from_str(VALID_P2SH_P2WSH_PROGRAM) {
+                Ok(witness_program) => {
+                    assert_eq!(0x00, witness_program.version);
+                    assert_eq!(vec![0x18, 0x63, 0x14, 0x3c, 0x14,
+                                    0xc5, 0x16, 0x68, 0x04, 0xbd,
+                                    0x19, 0x20, 0x33, 0x56, 0xda,
+                                    0x13, 0x6c, 0x98, 0x56, 0x78,
+                                    0xcd, 0x4d, 0x27, 0xa1, 0xb8,
+                                    0xc6, 0x32, 0x96, 0x04, 0x90,
+                                    0x32, 0x62
+                    ], witness_program.program);
+                }
+                Err(e) => assert_eq!(None, Some(e))
+            }
+        }
+
+        const VALID_OP_1_PROGRAM: &str = "0128751e76e8199196d454941c45d1b3a323f1433bd6751e76e8199196d454941c45d1b3a323f1433bd6";
+        const INVALID_OP_1_PROGRAM_LENGTH: &str = "0128751e76e8199196d454941c45d1b3a323f1433bd6751e76e8199196d454941c45d1b3a323f143";
+
+        #[test]
+        fn from_str_invalid_op_1_program_len() {
+            match WitnessProgram::from_str(INVALID_OP_1_PROGRAM_LENGTH) {
+                Ok(_) => assert_eq!(None, Some("Error, invalid program decoded successfully")),
```

Comment:
This case should not be reachable

Reply:
See how to test for errors here:
https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L845&L875